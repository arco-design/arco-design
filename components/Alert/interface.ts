import type { ReactNode } from 'react';

/**
 * @title Alert
 */
export interface AlertProps
  extends Omit<React.HTMLAttributes<HTMLDivElement>, 'className' | 'title' | 'content'> {
  /**
   * @zh 自定义类名
   * @en Custom className
   */
  className?: string | string[];
  /**
   * @zh 自定义操作项
   * @en Custom action item
   */
  action?: ReactNode;
  /**
   * @zh 是否可关闭
   * @en Whether Alert can be closed
   */
  closable?: boolean;
  closeable?: boolean; // typo
  /**
   * @zh 关闭的回调
   * @en Callback when Alert is closed
   */
  onClose?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  /**
   * @zh 关闭动画结束后执行的回调
   * @en Callback when Alert close animation is complete
   */
  afterClose?: () => void;
  /**
   * @zh 警告的类型
   * @en Type of Alert
   * @defaultValue info
   */
  type?: 'info' | 'success' | 'warning' | 'error';
  /**
   * @zh 标题
   * @en Alert title
   */
  title?: ReactNode;
  /**
   * @zh 内容
   * @en Alert content
   */
  content?: ReactNode;
  /**
   * @zh 可以指定自定义图标，`showIcon` 为 `true` 时生效。
   * @en Custom icon, effective when `showIcon` is `true`
   */
  icon?: ReactNode;
  /**
   * @zh 自定义关闭按钮
   * @en Custom close button
   */
  closeElement?: ReactNode;
  /**
   * @zh 是否显示图标
   * @en Whether to show icon
   * @defaultValue true
   */
  showIcon?: boolean;
  /**
   * @zh 是否用作顶部公告
   * @en Whether to show as banner
   */
  banner?: boolean;
}
