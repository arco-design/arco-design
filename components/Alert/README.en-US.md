`````
Component / Feedback

# Alert

Used to display warning information in a way that attracts attention.
`````

%%Content%%

## API

### Alert

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|banner|Whether to show as banner|boolean |`-`|
|closable|Whether Alert can be closed|boolean |`-`|
|showIcon|Whether to show icon|boolean |`true`|
|type|Type of Alert|'info' \| 'success' \| 'warning' \| 'error' |`info`|
|action|Custom action item|ReactNode |`-`|
|closeElement|Custom close button|ReactNode |`-`|
|content|Alert content|ReactNode |`-`|
|icon|Custom icon, effective when `showIcon` is `true`|ReactNode |`-`|
|title|Alert title|ReactNode |`-`|
|className|Additional css class|string \| string[] |`-`|
|style|Additional style|CSSProperties |`-`|
|afterClose|Callback when Alert close animation is complete|() => void |`-`|
|onClose|Callback when Alert is closed|(e) => void |`-`|
