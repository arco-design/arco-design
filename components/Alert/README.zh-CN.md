`````
组件 / 反馈

# 警告提示 Alert

向用户显示警告的信息时，通过警告提示，展现需要关注的信息。
`````

%%Content%%

## API

### Alert

|参数名|描述|类型|默认值|
|---|---|---|---|
|banner|是否用作顶部公告|boolean |`-`|
|closable|是否可关闭|boolean |`-`|
|showIcon|是否显示图标|boolean |`true`|
|type|警告的类型|'info' \| 'success' \| 'warning' \| 'error' |`info`|
|action|自定义操作项|ReactNode |`-`|
|closeElement|自定义关闭按钮|ReactNode |`-`|
|content|内容|ReactNode |`-`|
|icon|可以指定自定义图标，`showIcon` 为 `true` 时生效。|ReactNode |`-`|
|title|标题|ReactNode |`-`|
|className|节点类名|string \| string[] |`-`|
|style|节点样式|CSSProperties |`-`|
|afterClose|关闭动画结束后执行的回调|() => void |`-`|
|onClose|关闭的回调|(e) => void |`-`|
