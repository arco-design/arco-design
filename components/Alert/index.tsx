import React, { useState, useContext, ReactNode, forwardRef } from 'react';
import { CSSTransition } from 'react-transition-group';
import IconCheckCircleFill from '../../icon/react-icon/IconCheckCircleFill';
import IconCloseCircleFill from '../../icon/react-icon/IconCloseCircleFill';
import IconInfoCircleFill from '../../icon/react-icon/IconInfoCircleFill';
import IconExclamationCircleFill from '../../icon/react-icon/IconExclamationCircleFill';
import IconClose from '../../icon/react-icon/IconClose';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { AlertProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const defaultProps: AlertProps = {
  showIcon: true,
  type: 'info',
};

function Alert(baseProps: AlertProps, ref) {
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<AlertProps>(baseProps, defaultProps, componentConfig?.Alert);
  const {
    style,
    className,
    action,
    type = 'info',
    title,
    content,
    icon,
    showIcon,
    closable,
    closeable,
    afterClose,
    onClose,
    closeElement,
    banner,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('alert');
  const [visible, setVisible] = useState<boolean>(true);

  function renderIcon(type: string | void): ReactNode | null {
    if (icon) {
      return icon;
    }
    switch (type) {
      case 'info':
        return <IconInfoCircleFill />;
      case 'success':
        return <IconCheckCircleFill />;
      case 'warning':
        return <IconExclamationCircleFill />;
      case 'error':
        return <IconCloseCircleFill />;
      default:
        return null;
    }
  }

  function onHandleClose(e) {
    setVisible(false);
    onClose && onClose(e);
  }

  const classNames = cs(
    prefixCls,
    `${prefixCls}-${type}`,
    {
      [`${prefixCls}-with-title`]: title,
      [`${prefixCls}-banner`]: banner,
      [`${prefixCls}-rtl`]: rtl,
    },
    className
  );
  const _closable = 'closeable' in props ? closeable : closable;

  return (
    <CSSTransition
      in={visible}
      timeout={300}
      classNames="zoomInTop"
      unmountOnExit
      onExited={() => {
        afterClose && afterClose();
      }}
    >
      <div ref={ref} style={style} className={classNames} role="alert" {...rest}>
        {showIcon && <div className={`${prefixCls}-icon-wrapper`}>{renderIcon(type)}</div>}
        <div className={`${prefixCls}-content-wrapper`}>
          {title && <div className={`${prefixCls}-title`}>{title}</div>}
          {content && <div className={`${prefixCls}-content`}>{content}</div>}
        </div>
        {action && <div className={`${prefixCls}-action`}>{action}</div>}
        {_closable && (
          <button type="button" onClick={onHandleClose} className={`${prefixCls}-close-btn`}>
            {closeElement || <IconClose />}
          </button>
        )}
      </div>
    </CSSTransition>
  );
}

const AlertComponent = forwardRef<unknown, AlertProps>(Alert);

AlertComponent.displayName = 'Alert';

export default AlertComponent;

export { AlertProps };
