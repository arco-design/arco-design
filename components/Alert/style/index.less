@import './token.less';

@alert-prefix-cls: ~'@{prefix}-alert';

.@{alert-prefix-cls} {
  box-sizing: border-box;
  border-radius: @alert-border-radius;
  padding: (@alert-padding-vertical - @alert-border-width)
    (@alert-padding-horizontal - @alert-border-width);
  font-size: @alert-font-size-text-content;
  overflow: hidden;
  display: flex;
  width: 100%;
  text-align: left;
  align-items: center;
  line-height: @alert-line-height;

  &-with-title {
    padding: (@alert-padding-vertical_with_title - @alert-border-width)
      (@alert-padding-horizontal_with_title - @alert-border-width);
  }

  &-with-title {
    align-items: flex-start;
  }

  &-info {
    border: @alert-border-width solid @alert-info-color-border;
    background-color: @alert-info-color-bg;
  }

  &-success {
    border: @alert-border-width solid @alert-success-color-border;
    background-color: @alert-success-color-bg;
  }

  &-warning {
    border: @alert-border-width solid @alert-warning-color-border;
    background-color: @alert-warning-color-bg;
  }

  &-error {
    border: @alert-border-width solid @alert-error-color-border;
    background-color: @alert-error-color-bg;
  }

  &-banner {
    border: none;
    border-radius: 0;
  }

  &-content-wrapper {
    position: relative;
    flex: 1;
  }

  &-title {
    font-size: @alert-font-size-text-title;
    font-weight: @alert-font-weight-title;
    line-height: @alert-title-line-height;
    margin-bottom: @alert-title-margin-bottom;
  }

  &-info &-title {
    color: @alert-info-color-text-title;
  }

  &-info &-content {
    color: @alert-info-color-text-content;
  }

  &-info&-with-title &-content {
    color: @alert-info-color-text-content_title;
  }

  &-success &-title {
    color: @alert-success-color-text-title;
  }

  &-success &-content {
    color: @alert-success-color-text-content;
  }

  &-success&-with-title &-content {
    color: @alert-success-color-text-content_title;
  }

  &-warning &-title {
    color: @alert-warning-color-text-title;
  }

  &-warning &-content {
    color: @alert-warning-color-text-content;
  }

  &-warning&-with-title &-content {
    color: @alert-warning-color-text-content_title;
  }

  &-error &-title {
    color: @alert-error-color-text-title;
  }

  &-error &-content {
    color: @alert-error-color-text-content;
  }

  &-error&-with-title &-content {
    color: @alert-error-color-text-content_title;
  }

  &-icon-wrapper {
    margin-right: @alert-margin-icon-right;
    height: @alert-line-height * @alert-font-size-text-content;
    display: flex;
    align-items: center;

    svg {
      font-size: @alert-font-size-icon;
    }
  }

  &-with-title &-icon-wrapper {
    height: @alert-title-line-height * @alert-font-size-text-title;
  }

  &-with-title &-icon-wrapper svg {
    font-size: @alert-font-size-icon_with_title;
  }

  &-info &-icon-wrapper svg {
    color: @alert-info-color-icon;
  }

  &-success &-icon-wrapper svg {
    color: @alert-success-color-icon;
  }

  &-warning &-icon-wrapper svg {
    color: @alert-warning-color-icon;
  }

  &-error &-icon-wrapper svg {
    color: @alert-error-color-icon;
  }

  &-close-btn {
    box-sizing: border-box;
    padding: 0;
    border: none;
    outline: none;
    font-size: @alert-font-size-close-icon;
    color: @alert-color-close-icon;
    background-color: transparent;
    cursor: pointer;
    transition: color @transition-duration-1 @transition-timing-function-linear;
    margin-left: @alert-margin-close-icon-left;
    top: 4px;
    right: 0;

    &:hover {
      color: @alert-color-close-icon_hover;
    }
  }

  &-action + &-close-btn {
    margin-left: @alert-margin-action-right;
  }

  &-action {
    margin-left: @alert-margin-action-left;
  }

  &-with-title &-close-btn {
    margin-top: 0;
    margin-right: 0;
  }
}

.@{alert-prefix-cls}-rtl {
  direction: rtl;
  text-align: right;

  .@{alert-prefix-cls}-with-title {
    align-items: flex-end;
  }

  .@{alert-prefix-cls}-icon-wrapper {
    margin-right: 0;
    margin-left: @alert-margin-icon-right;
  }

  .@{alert-prefix-cls}-close-btn {
    right: initial;
    left: 0;
  }

  .@{alert-prefix-cls}-action {
    margin-left: 0;
    margin-right: @alert-margin-action-left;

    + .@{alert-prefix-cls}-close-btn {
      margin-left: 0;
      margin-right: @alert-margin-action-right;
    }
  }
}
