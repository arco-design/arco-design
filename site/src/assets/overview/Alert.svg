<svg width="128" height="128" viewBox="0 0 128 128" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="108" height="24" transform="translate(10 52)" fill="#E8F3FF"/>
<rect x="18" y="60" width="8" height="8" rx="4" fill="#6AA1FF"/>
<rect x="30" y="62" width="80" height="4" rx="2" fill="#BEDAFF"/>
</svg>
