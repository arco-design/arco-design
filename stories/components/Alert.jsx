import React, { Component } from 'react';
import { Alert } from '@self';
import { IconBug } from '@self/icon';

class Demo extends Component {
  constructor(props) {
    super(props);

    this.state = {};
  }

  render() {
    return (
      <>
        <Alert
          showIcon
          type="info"
          title="Info"
          content="ContentContentContentContentContentContentContentContentContentContentContentContentContentContentContentContentContent"
          style={{ marginTop: 10 }}
        />
        <Alert
          showIcon
          type="success"
          title="Success"
          content="ContentContentContentContentContentContentContentContentContentContentContentContentContentContentContentContentContent~"
          style={{ marginTop: 10 }}
        />
        <Alert
          showIcon
          type="warning"
          title="Warning"
          content="Content~"
          style={{ marginTop: 10 }}
        />
        <Alert showIcon type="error" title="Error" content="Content~" style={{ marginTop: 10 }} />
        <Alert
          icon={<IconBug style={{ color: 'green' }} />}
          type="normal"
          title="Normal"
          content="Content~"
          style={{ marginTop: 10 }}
        />
        <Alert
          type="success"
          title="没有图标"
          content="ContentContentContentContentContentContentContentContentContentContentContentContentContentContentContentContentContent~"
          style={{ marginTop: 10 }}
        />
      </>
    );
  }
}

export default Demo;
