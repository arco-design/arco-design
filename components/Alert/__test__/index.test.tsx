import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Alert from '..';

mountTest(Alert);
componentConfigTest(Alert, 'Alert');

describe('Alert', () => {
  it('showIcon correctly', () => {
    const alert = mount(<Alert type="error" title="Error" content="Content~" />);

    expect(alert.find('.arco-icon-close-circle-fill').length).toBe(1);
  });

  it('onClose and afterClose', () => {
    const closeFn = jest.fn();
    const afterCloseFn = jest.fn();
    const alert = mount(
      <Alert
        closable
        onClose={closeFn}
        afterClose={afterCloseFn}
        type="info"
        title="Title"
        content="Content"
      />
    );

    jest.useFakeTimers();
    alert.find('.arco-alert-close-btn').at(0).simulate('click');

    expect(closeFn.mock.calls.length).toBe(1);
    jest.runAllTimers();
    expect(afterCloseFn.mock.calls.length).toBe(1);
  });
});
