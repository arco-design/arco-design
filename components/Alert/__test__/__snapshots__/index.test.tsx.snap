// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Alert default 1`] = `
<div
  class="arco-alert arco-alert-info"
  role="alert"
>
  <div
    class="arco-alert-icon-wrapper"
  >
    <svg
      aria-hidden="true"
      class="arco-icon arco-icon-info-circle-fill"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="4"
      viewBox="0 0 48 48"
    >
      <path
        clip-rule="evenodd"
        d="M24 44c11.046 0 20-8.954 20-20S35.046 4 24 4 4 12.954 4 24s8.954 20 20 20Zm2-30a1 1 0 0 0-1-1h-2a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1v-2Zm0 17h1a1 1 0 0 1 1 1v2a1 1 0 0 1-1 1h-6a1 1 0 0 1-1-1v-2a1 1 0 0 1 1-1h1v-8a1 1 0 0 1-1-1v-2a1 1 0 0 1 1-1h3a1 1 0 0 1 1 1v11Z"
        fill="currentColor"
        fill-rule="evenodd"
        stroke="none"
      />
    </svg>
  </div>
  <div
    class="arco-alert-content-wrapper"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Alert set className globally 1`] = `
<div
  class="arco-alert arco-alert-info global-default-class"
  role="alert"
>
  <div
    class="arco-alert-icon-wrapper"
  >
    <svg
      aria-hidden="true"
      class="arco-icon arco-icon-info-circle-fill"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="4"
      viewBox="0 0 48 48"
    >
      <path
        clip-rule="evenodd"
        d="M24 44c11.046 0 20-8.954 20-20S35.046 4 24 4 4 12.954 4 24s8.954 20 20 20Zm2-30a1 1 0 0 0-1-1h-2a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1v-2Zm0 17h1a1 1 0 0 1 1 1v2a1 1 0 0 1-1 1h-6a1 1 0 0 1-1-1v-2a1 1 0 0 1 1-1h1v-8a1 1 0 0 1-1-1v-2a1 1 0 0 1 1-1h3a1 1 0 0 1 1 1v11Z"
        fill="currentColor"
        fill-rule="evenodd"
        stroke="none"
      />
    </svg>
  </div>
  <div
    class="arco-alert-content-wrapper"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Alert set className globally and component set className itself. 1`] = `
<div
  class="arco-alert arco-alert-info component-class"
  role="alert"
>
  <div
    class="arco-alert-icon-wrapper"
  >
    <svg
      aria-hidden="true"
      class="arco-icon arco-icon-info-circle-fill"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="4"
      viewBox="0 0 48 48"
    >
      <path
        clip-rule="evenodd"
        d="M24 44c11.046 0 20-8.954 20-20S35.046 4 24 4 4 12.954 4 24s8.954 20 20 20Zm2-30a1 1 0 0 0-1-1h-2a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1v-2Zm0 17h1a1 1 0 0 1 1 1v2a1 1 0 0 1-1 1h-6a1 1 0 0 1-1-1v-2a1 1 0 0 1 1-1h1v-8a1 1 0 0 1-1-1v-2a1 1 0 0 1 1-1h3a1 1 0 0 1 1 1v11Z"
        fill="currentColor"
        fill-rule="evenodd"
        stroke="none"
      />
    </svg>
  </div>
  <div
    class="arco-alert-content-wrapper"
  />
</div>
`;
